group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -0,0 +1,865 @@
+        ConsumerGroupMember member = group.getOrMaybeCreateMember(memberId, createIfNotExists);
+        throwIfMemberEpochIsInvalid(member, memberEpoch, ownedTopicPartitions);
+
+        if (memberEpoch == 0) {
+            log.info("[GroupId " + groupId + "] Member " + memberId + " re-joins the consumer group.");
+        }
+
+        // Update the subscription part of the member if we received new values. If the member has
+        // changed, we write it to the log. If the subscribed topics have changed, we also recompute
+        // the subscription metadata.
+        int groupEpoch = group.groupEpoch();
+        Map<String, TopicMetadata> subscriptionMetadata = group.subscriptionMetadata();
+        ConsumerGroupMember updatedMember = new ConsumerGroupMember.Builder(member)
+            .maybeUpdateInstanceId(Optional.ofNullable(instanceId))
+            .maybeUpdateRackId(Optional.ofNullable(rackId))
+            .maybeUpdateRebalanceTimeoutMs(ofSentinel(rebalanceTimeoutMs))
+            .maybeUpdateServerAssignorName(Optional.ofNullable(assignorName))
+            .maybeUpdateSubscribedTopicNames(Optional.ofNullable(subscribedTopicNames))
+            .maybeUpdateSubscribedTopicRegex(Optional.ofNullable(subscribedTopicRegex))
+            .setClientId(clientId)
+            .setClientHost(clientHost)
+            .build();
+
+        if (!updatedMember.equals(member)) {
+            records.add(newMemberSubscriptionRecord(groupId, updatedMember));
+
+            if (!updatedMember.subscribedTopicNames().equals(member.subscribedTopicNames())) {
+                log.info("[GroupId " + groupId + "] Member " + memberId + " updated its subscribed topics to: " +
+                    updatedMember.subscribedTopicNames());
+
+                subscriptionMetadata = group.computeSubscriptionMetadata(
+                    updatedMember.memberId(),
+                    updatedMember.subscribedTopicNames(),
+                    topicsImage
+                );
+
+                if (!subscriptionMetadata.equals(group.subscriptionMetadata())) {
+                    log.info("[GroupId " + groupId + "] Computed new subscription metadata: "
+                        + subscriptionMetadata + ".");
+                    records.add(newGroupSubscriptionMetadataRecord(groupId, subscriptionMetadata));
+                }
+
+                groupEpoch += 1;
+                records.add(newGroupEpochRecord(groupId, groupEpoch));
+
+                log.info("[GroupId " + groupId + "] Bumped group epoch to " + groupEpoch + ".");
+            }
+
+            member = updatedMember;
+        }
+
+        // Update target assignment if needed. If the new target has any changes, we write the
+        // changes to the log.
+        int targetAssignmentEpoch = group.assignmentEpoch();
+        Assignment targetAssignment = group.targetAssignment(memberId);
+        if (groupEpoch > targetAssignmentEpoch) {
```

Comment:
Is the incrementing on line 497 the only way groupEpoch will be greater than the targetAssignmentEpoch?

This is a little confusing to me since I would expect the target to be higher. I think saying generic "groupEpoch" also confuses me about the source of the data.

Reply:
> Is the incrementing on line 497 the only way groupEpoch will be greater than the targetAssignmentEpoch?

At the moment, this is the only way but other ways will come. I think that it is safe to trigger the computation whenever the group epoch is larger than the target assignment epoch.

> This is a little confusing to me since I would expect the target to be higher. I think saying generic "groupEpoch" also confuses me about the source of the data.

We bump the group epoch when the group has changed (e.g. new members, new subscriptions, etc.). Then, we update the target assignment when we detect that it is older than the group metadata. However, the member epoch should always be lower or equals to the target assignment epoch.

Regarding "groupEpoch", I am not sure to understand your point. Are you saying that the name is confusing?

Reply:
maybe it would be good to include a comment on the relationship between the group, target assignment, and member epochs.

Reply:
I think the comment 
> We bump the group epoch when the group has changed (e.g. new members, new subscriptions, etc.). 

makes things a little clearer. But I guess my confusion was it seems like this should always cause the assignment to change. I think I also got a bit confused because on earlier PRs I thought that the target assignment epoch was the epoch we get we at at when the assignment is complete. But maybe I am confused somewhere.

Reply:
> it seems like this should always cause the assignment to change
what's "this"?

> the target assignment epoch was the epoch we get we at at when the assignment is complete
that's right. the target assignment epoch is what every member will try to converge to by revoking/assigning partitions. once we build the target assignment, we bump the target assignment epoch (L532)

Reply:
When we bump the group epoch, it seems like we will also always reassign. So I guess I was just trying to figure out if there is a case where the group epoch is different from the assignment epoch.

Reply:
Following up here after re-reading this and the KIP. It seems that the group epoch is only different from the assignment epoch before we assign. Is this correct? I'm just trying to figure out if there is any other case. It also seems that we update the assignment epoch to be the group epoch when the assignment completes. This all happens within the method so I guess I'm confused why these need to be separate epoch fields. 

Is there a time this isn't the case? I can imagine errors with assignment, so is that the main reason we have separate values here?

I also found lines 519 confusing:
>  int targetAssignmentEpoch = group.assignmentEpoch();

Here "targetAssignmentEpoch" is just current epoch of the assignment. We end up updating it to the groupEpoch after the assignment is done. I think the usage of the word target is a bit confusing here. I guess this epoch is the target for the member epochs to get to?

Reply:
Rereading this and the KIP -- is the only time we have a different assignment epoch from the group epoch is when assignment fails? I also assume that other members should learn about this assignment. I think I'm missing how that is done. I don't see the group values being set anywhere.

Reply:
Yeah, let me try to re-explain it with my words.

Basically, we have three epochs: the group epoch, the assignment epoch, and each member has its own epoch.
- When the group changes (e.g. new member, updated subscription, etc), the group epoch is bumped in order to note that the topology of the group has changed.
- When the assignment epoch is smaller than the group epoch, it means that the assignment for the group is stable so we need to recompute it. We generate a new assignment with the epoch matching the group epoch in order to note that the assignment is for the latest group topology.
- When a member has a smaller epoch than the assignment epoch, it means that its assignment is stale so we need to reconcile it to converge it to that epoch.

All the epochs are equal when the group is stable. This means that all the members have converged to the desired assignment.

> Rereading this and the KIP -- is the only time we have a different assignment epoch from the group epoch is when assignment fails? I also assume that other members should learn about this assignment. I think I'm missing how that is done -- is this through the records methods?
So for the next heartbeat request of another member, line 552 would see updatedMember.nextMemberEpoch() != targetAssignmentEpoch as true since the group epoch and assignment epoch would be equal, but the member epoch would still be behind?

Yeah, we could have a difference between the assignment epoch and the group epoch when the assignment is not computed immediately. In the current implementation, we almost never have this but keep in mind that we will support client side assignors. In this case, the computation will take some time so we will see it more.

The member learn about their assignment when they heartbeat. This is when we do the reconciliation process. We reconcile a member in two cases: 1) the member is not stable; 2) the target assignment has changed since the last time we reconciled the member. This is done [here](https://github.com/apache/kafka/pull/13639/files#diff-00f0f81cf13e66781777d94f7d2e68a581663385c37e98792507f2294c91bb09R569).